<a href="https://colab.research.google.com/github/cristiandarioortegayubro/UNI/blob/main/Presentaci%C3%B3n2023_M3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://github.com/cristiandarioortegayubro/UNI/blob/main/UNI%2039%20Aniversario.jpg?raw=true)


<p align="center">
<img src="https://github.com/cristiandarioortegayubro/BDS/blob/main/images/Logo%20Scikit-learn.png?raw=true">
</p>


 # **<font color="DarkBlue">Análisis de datos y Machine Learning con Python</font>**

 # **<font color="DarkBlue">Instructores</font>**

<table border="1">
  <tr>
    <td>
    <h3 align="center"><font color="DarkBlue">Gustavo Raul Machin Urbay</font></h3>
    </td>
    <td>
    <h3 align="center"><font color="DarkBlue">Cristian Dario Ortega Yubro</font></h3>
    </td>
  </tr>

  <tr>
    <td align="center">
    <img src="https://media.licdn.com/dms/image/C4E03AQHVW8FGqKW5TQ/profile-displayphoto-shrink_200_200/0/1642528082273?e=1698278400&v=beta&t=8pXiMbD4klPgD_Y_sFg1NHH12ANHKXr2YGGluppmFmE">        
    </td>
    <td align="center">
    <img src="https://media.licdn.com/dms/image/D4D03AQGQvYNSGkVLOQ/profile-displayphoto-shrink_200_200/0/1674752467497?e=1698278400&v=beta&t=25lFdtPMJMhBN4RVI9ja9oESMgu8kk_8OANDkdztB6w">
    </td>
  </tr>

  <tr>
    <td>
    <a href="https://www.linkedin.com/in/gustavomachinurbay/">https://www.linkedin.com/in/gustavomachinurbay</a>
    </td>
    <td>
    <a href="https://www.linkedin.com/in/cristiandarioortegayubro/">https://www.linkedin.com/in/cristiandarioortegayubro</a>
    </td>
  </tr>
</table>


 # **<font color="DarkBlue">Contenidos</font>**

 ## **<font color="DarkBlue"> 1 - Preprocesamiento avanzado con Scikit-learn</font>**

✅ escalamiento de variables,
<br>
✅ extracción de variables,
<br>
✅ manejo de datos desbalanceados,
<br>
✅ codificación de variables categóricas


   


 ## **<font color="DarkBlue"> 2 - Pipeline con Scikit-learn I</font>**

✅ tratamiento de variables numéricas,
<br>
✅ tratamiento de variables categóricas,
<br>
✅ transformadores de variables numéricas y categóricas
  
   


 ## **<font color="DarkBlue"> 3 - Pipeline con Scikit-learn II</font>**

✅ división del conjunto de datos de entrenamiento y prueba,
<br>
✅ validación cruzada





 ## **<font color="DarkBlue"> 4 - Pipeline con Scikit-learn III</font>**

✅ entrenamiento,
<br>
✅ predicción y evaluación de modelos,
<br>
✅ pipeline transformadores + pipeline modelo

   

 ## **<font color="DarkBlue"> 5 - Modelos de Regresión</font>**

✅ regresión lineal,
<br>
✅ árbol de regresión,
<br>
✅ evaluación y selección de modelos de regresión



 ## **<font color="DarkBlue"> 6 - Modelos de Clasificación</font>**

✅ clasificación binaria y multiclase,
<br>
✅ regresión logística,
<br>
✅ árbol de clasificación,
<br>
✅ evaluación y selección de modelos de clasificación




 ## **<font color="DarkBlue"> 7 - Modelos de Clustering</font>**

✅ k-means,
<br>
✅ evaluación de la calidad de los clusters



 ## **<font color="DarkBlue"> 8 - Métricas</font>**

✅ accuracy,
<br>
✅ precision,
<br>
✅ recall,
<br>
✅ f1-score,
<br>
✅ matriz de confusión,
<br>
✅ curva ROC y área debajo de la curva (AUC)




 # **<font color="DarkBlue"> 💗 Usando Pandas, Plotly y Scikit-learn...</font>**

 ## **<font color="DarkBlue">Carga de bibliotecas necesarias</font>**

In [ ]:
import pandas as pd
import plotly.express as px
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import make_pipeline

Las bibliotecas que utilizamos, en el código anterior:

1. `import pandas as pd`: En esta línea, estás importando la biblioteca Pandas y la estás nombrando como "pd". Pandas es una biblioteca ampliamente utilizada para el análisis de datos en Python. Proporciona estructuras de datos y funciones que facilitan la manipulación y el análisis de datos, especialmente para datos tabulares.

2. `import plotly.express as px`: Aquí estás importando un módulo específico llamado "express" de la biblioteca Plotly y lo estás nombrando como "px". Plotly es una biblioteca que se utiliza para crear visualizaciones interactivas en Python. El módulo "express" proporciona una interfaz simple y rápida para crear gráficos y visualizaciones de datos.

3. `from sklearn.preprocessing import StandardScaler`: Estás importando la clase "StandardScaler" del módulo "preprocessing" de la biblioteca scikit-learn (sklearn). El escalado estándar es una técnica común en el preprocesamiento de datos en la que se transforman los datos para que tengan una media de 0 y una desviación estándar de 1. Esto es útil para algoritmos de aprendizaje automático que son sensibles a la escala de las características.

4. `from sklearn.model_selection import train_test_split`: Aquí estás importando la función "train_test_split" del módulo "model_selection" de sklearn. Esta función se utiliza para dividir un conjunto de datos en dos partes: un conjunto de entrenamiento y un conjunto de prueba. Esto es fundamental para evaluar el rendimiento de los modelos de aprendizaje automático en datos no vistos.

5. `from sklearn.linear_model import LinearRegression`: Estás importando la clase "LinearRegression" del módulo "linear_model" de sklearn. La regresión lineal es un algoritmo de aprendizaje automático utilizado para modelar la relación entre una variable dependiente y una o más variables independientes. Es especialmente útil cuando se busca una relación lineal entre las variables.

6. `from sklearn.pipeline import make_pipeline`: Aquí estás importando la función "make_pipeline" del módulo "pipeline" de sklearn. Los tuberías (pipelines) son una forma eficiente de encadenar varios pasos en un flujo de trabajo de aprendizaje automático. La función "make_pipeline" crea una tubería en la que los datos pasan secuencialmente a través de una serie de pasos, como el preprocesamiento y la estimación del modelo.

Estas líneas de código son esenciales para construir y entrenar modelos de aprendizaje automático utilizando la biblioteca scikit-learn y para realizar análisis y visualizaciones de datos con Pandas y Plotly.

## **<font color="DarkBlue">Conjunto de Datos**

In [ ]:
url = "https://raw.githubusercontent.com/cristiandarioortegayubro/BDS/main/datasets/Ecommerce_Customers.csv"
datos = pd.read_csv(url)

<p align="justify">
Es una empresa de comercio electrónico con sede en la ciudad de Nueva York que vende ropa en línea, pero también tienen sesiones de asesoramiento sobre estilo y ropa en la tienda. Los clientes que vienen a la tienda, tienen sesiones/reuniones con un estilista personal, luego pueden ordenar a través de una aplicación móvil o sitio web la ropa que desean.
<br>
<br>
Intentan identificar si debe enfocarse en mejorar su experiencia de aplicación movil o en su página web.
<br>
<br>
Este es un conjunto de datos de los clientes de la empresa. El cual tiene información del cliente, como correo electrónico, dirección postal y su color de avatar. También tiene columnas de valores numéricos:

* **Avg. Session Length**: duración promedio de sesiones de asesoramiento de estilo en la tienda.

* **Time on App**: tiempo promedio de permanencia en la aplicación en minutos.

* **Time on Website**: tiempo promedio de permanencia en el sitio web en minutos.

* **Length of Membership**: cuántos años ha sido miembro el cliente.

* **Yearly Amount Spent**: monto anual gastado en la plataforma.


## **<font color="DarkBlue">Análisis y Transformación de Datos**

In [ ]:
datos = datos.drop(["Avatar","Email","Address"], axis=1)

<p align="justify">
En esta línea de código, se está utilizando el método `drop()` de Pandas para eliminar columnas específicas de un DataFrame llamado "datos". Aquí hay una explicación paso a paso:

- `datos`: Este es el nombre del DataFrame en el que estás trabajando. Supongo que previamente has cargado o creado un DataFrame llamado "datos".

- `.drop()`: Este es el método de Pandas que se utiliza para eliminar elementos (filas o columnas) de un DataFrame.

- `["Avatar", "Email", "Address"]`: Aquí se proporciona una lista de etiquetas de columnas que deseas eliminar. En este caso, las columnas que se eliminarán son "Avatar", "Email" y "Address".

- `axis=1`: Este argumento se utiliza para indicar que estás eliminando columnas en lugar de filas. El valor `axis=1` se refiere al eje horizontal (columnas), mientras que `axis=0` se refiere al eje vertical (filas).

<p align="justify">
En resumen, la línea de código está eliminando las columnas "Avatar", "Email" y "Address" del DataFrame "datos". Esto puede ser útil cuando deseas eliminar características que no son relevantes para tu análisis o modelado de datos.

In [ ]:
# Renombramos columnas:
datos.rename(columns={"Avg. Session Length": "Tiempo_sesión",
                      "Time on App":"Tiempo_app",
                      "Time on Website":"Tiempo_web",
                      "Length of Membership":"Años_miembro",
                      "Yearly Amount Spent":"Gasto_anual"},inplace=True)

<p align="justify">
Este fragmento de código se utiliza para cambiar los nombres de las columnas en un DataFrame de Pandas. A continuación, definimos lo que está sucediendo:

- `datos`: Este es el DataFrame en el que deseas renombrar las columnas.

- `.rename(columns={...}, inplace=True)`: Este es el método que se utiliza para renombrar las columnas. Se pasa un diccionario como argumento para especificar los cambios de nombres de las columnas. El parámetro `inplace=True` indica que los cambios deben realizarse directamente en el DataFrame "datos", sin necesidad de asignar el resultado a una nueva variable.

Dentro del diccionario proporcionado, las claves son los nombres de las columnas que deseas cambiar, y los valores son los nuevos nombres que deseas asignarles. Aquí está el mapeo que se está haciendo en este caso:

- `"Avg. Session Length"` se renombra como `"Tiempo_sesión"`
- `"Time on App"` se renombra como `"Tiempo_app"`
- `"Time on Website"` se renombra como `"Tiempo_web"`
- `"Length of Membership"` se renombra como `"Años_miembro"`
- `"Yearly Amount Spent"` se renombra como `"Gasto_anual"`

<p align="justify">
En resumen, este código cambia los nombres de las columnas en el DataFrame "datos" según el mapeo especificado en el diccionario y lo hace directamente en el mismo DataFrame debido al parámetro `inplace=True`.

## **<font color="DarkBlue">Matriz de Correlación**

In [ ]:
corr_matrix = round(datos.corr(),3)
px.imshow(corr_matrix,
          title = "Matriz de Correlacion",
          text_auto=True,
          labels={"color":"Indice"})

<p align="justify">
Este código se utiliza para calcular y visualizar una matriz de correlación de las variables en un DataFrame utilizando la biblioteca Plotly Express. A continuación, se detalla lo que está sucediendo:

- `corr_matrix = round(datos.corr(), 3)`: En esta línea, estás calculando la matriz de correlación de las variables en el DataFrame "datos" utilizando el método `.corr()` de Pandas. La función `round()` se utiliza para redondear los valores de la matriz de correlación a tres decimales. La matriz de correlación muestra cómo las variables numéricas en el DataFrame se relacionan entre sí, con valores que van desde -1 (correlación negativa perfecta) hasta 1 (correlación positiva perfecta).

- `px.imshow(corr_matrix, title="Matriz de Correlacion", text_auto=True, labels={"color": "Indice"})`: Aquí estás utilizando Plotly Express para crear una visualización de la matriz de correlación. `px.imshow()` se utiliza para crear un mapa de calor a partir de los datos proporcionados. Los argumentos son los siguientes:
  - `corr_matrix`: Es la matriz de correlación que calculaste previamente.
  - `title="Matriz de Correlacion"`: Establece el título de la visualización.
  - `text_auto=True`: Esto hace que los valores de correlación se muestren automáticamente en cada celda del mapa de calor.
  - `labels={"color": "Indice"}`: Etiqueta el eje de color del mapa de calor como "Índice", lo que indica que representa los valores de correlación.

<p align="justify">
En resumen, este código te permite calcular y visualizar de manera gráfica la matriz de correlación entre las variables numéricas en el DataFrame "datos". La visualización muestra patrones de correlación entre las variables, lo que puede ayudarte a comprender cómo se relacionan entre sí y a tomar decisiones informadas en análisis de datos y modelado.

<p align="justify">
A primera vista, se evidencia una relación fuerte entre la variable explicativa <code>Años_miembro</code> y la variable <code>Gasto_anual</code>. Asimismo, se puede apreciar una relación moderada entre la variable explicativa <code>Tiempo_app</code> y la variable <code>Gasto_anual</code>. La relación entre <code>Tiempo_web</code> y <code>Gasto_anual</code> es nula.
<br>
<br>
No obstante, no se observa una correlación significativa entre las variables explicativas entre sí. Esto indica que las variables explicativas son <b>independientes</b> entre sí y no hay una relación directa o sistemática entre ellas en el conjunto de datos analizado. Cada una de ellas puede proporcionar información única y no redundante para explicar la variable respuesta (<code>Gasto_anual</code>)

## **<font color="DarkBlue">Gráficos de Dispersión de cada variable explicativa**

In [ ]:
X = datos[['Tiempo_sesión', 'Tiempo_app','Tiempo_web','Años_miembro']]
y = datos['Gasto_anual']

In [ ]:
for i in X.columns:
  fig = px.scatter(datos,
             x = i,
             y = y,
             trendline="ols",
             trendline_color_override="orange",
             template = "plotly_dark",
             title = f"La variable {i}")
  fig.show()

<p align="justify">
Este código se utiliza para calcular y visualizar una matriz de correlación de las variables en un DataFrame utilizando la biblioteca Plotly Express. A continuación, detallamos lo que está sucediendo:

- `corr_matrix = round(datos.corr(), 3)`: En esta línea, estás calculando la matriz de correlación de las variables en el DataFrame "datos" utilizando el método `.corr()` de Pandas. La función `round()` se utiliza para redondear los valores de la matriz de correlación a tres decimales. La matriz de correlación muestra cómo las variables numéricas en el DataFrame se relacionan entre sí, con valores que van desde -1 (correlación negativa perfecta) hasta 1 (correlación positiva perfecta).

- `px.imshow(corr_matrix, title="Matriz de Correlacion", text_auto=True, labels={"color": "Indice"})`: Aquí estás utilizando Plotly Express para crear una visualización de la matriz de correlación. `px.imshow()` se utiliza para crear un mapa de calor a partir de los datos proporcionados. Los argumentos son los siguientes:
  - `corr_matrix`: Es la matriz de correlación que calculaste previamente.
  - `title="Matriz de Correlacion"`: Establece el título de la visualización.
  - `text_auto=True`: Esto hace que los valores de correlación se muestren automáticamente en cada celda del mapa de calor.
  - `labels={"color": "Indice"}`: Etiqueta el eje de color del mapa de calor como "Índice", lo que indica que representa los valores de correlación.

<p align="justify">
En resumen, este código te permite calcular y visualizar de manera gráfica la matriz de correlación entre las variables numéricas en el DataFrame "datos". La visualización muestra patrones de correlación entre las variables, lo que puede ayudarte a comprender cómo se relacionan entre sí y a tomar decisiones informadas en análisis de datos y modelado.

 ## **<font color="DarkBlue">Estandarización de variables numéricas</font>**

<p align="justify">
<code>StandardScaler</code> es una clase de la biblioteca <code>scikit-learn</code> en Python que se utiliza para estandarizar variables numéricas en un conjunto de datos. La estandarización es un paso común en el preprocesamiento de datos antes de aplicar técnicas de aprendizaje automático, ya que ayuda a que las variables tengan una escala común y elimina cualquier sesgo relacionado con la escala de las variables explicativas.


In [ ]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [ ]:
X_scaled = pd.DataFrame(X_scaled, columns = X.columns)
X_scaled.head()

Tiempo_sesión  Tiempo_app  Tiempo_web  Años_miembro
0       1.456351    0.607280    2.493589      0.550107
1      -1.136502   -0.949464    0.206556     -0.870927
2      -0.052723   -0.727139    0.049681      0.572067
3       1.263010    1.676390   -0.335978     -0.413996
4       0.279838    0.747770    0.471737      0.914422

<p align="justify">
Este fragmento de código se utiliza para estandarizar las variables independientes en el DataFrame "X" utilizando la clase `StandardScaler` de scikit-learn (sklearn). Aquí está el desglose:

1. `scaler = StandardScaler()`: Creas una instancia del objeto `StandardScaler`, que se utilizará para estandarizar los datos. El escalado estándar transforma los datos para que tengan una media de 0 y una desviación estándar de 1, lo que es útil para muchos algoritmos de aprendizaje automático.

2. `X_scaled = scaler.fit_transform(X)`: Utilizas el método `fit_transform()` del escalador en el DataFrame "X". Esto calcula las medias y desviaciones estándar de las características en "X" y luego transforma los datos para que estén escalados. El resultado es un conjunto de datos escalados, almacenado en la variable "X_scaled".

3. `X_scaled = pd.DataFrame(X_scaled, columns=X.columns)`: Creas un nuevo DataFrame llamado "X_scaled" a partir de los datos escalados. Utilizas la clase `pd.DataFrame()` de Pandas para crear un DataFrame con los datos escalados. También asignas los nombres de las columnas originales de "X" a las columnas del nuevo DataFrame "X_scaled".

4. `X_scaled.head()`: Finalmente, utilizas el método `.head()` en el DataFrame "X_scaled" para mostrar las primeras filas de los datos escalados. Esto te proporciona una vista previa de cómo se ven los datos después de aplicar el escalado estándar.

<p align="justify">
En resumen, este código está realizando la estandarización de las variables independientes en el DataFrame "X". Después de escalar los datos, se crea un nuevo DataFrame "X_scaled" con los datos escalados y se muestra una vista previa de las primeras filas para verificar el resultado del proceso de escalado.

 ## **<font color="DarkBlue">División del conjunto de datos</font>**

In [ ]:
X_train, X_test, y_train, y_test  = train_test_split(X_scaled,
                                                     y,
                                                     random_state=123)

<p align="justify">
Este fragmento de código se utiliza para dividir los conjuntos de datos escalados en conjuntos de entrenamiento y prueba utilizando la función `train_test_split` de scikit-learn (sklearn). Aquí está el desglose:

- `X_scaled`: Este es el DataFrame de características escaladas que has preparado previamente.

- `y`: Esta es la Serie de la variable dependiente que estás tratando de predecir.

- `train_test_split(X_scaled, y, random_state=123)`: Aquí estás utilizando la función `train_test_split` para dividir tus datos en conjuntos de entrenamiento y prueba. Los argumentos son los siguientes:
  - `X_scaled`: El DataFrame de características escaladas.
  - `y`: La Serie de la variable dependiente.
  - `random_state=123`: Este argumento establece una semilla para la generación de números aleatorios. Proporcionar un valor específico a `random_state` asegura que la división de datos sea reproducible, lo que significa que obtendrás la misma división cada vez que ejecutes el código con la misma semilla.

- `X_train, X_test, y_train, y_test`: Las variables que preceden a la asignación almacenarán los conjuntos resultantes después de la división:
  - `X_train`: Contiene el conjunto de entrenamiento de características.
  - `X_test`: Contiene el conjunto de prueba de características.
  - `y_train`: Contiene el conjunto de entrenamiento de la variable dependiente.
  - `y_test`: Contiene el conjunto de prueba de la variable dependiente.

<p align="justify">
En resumen, este código realiza una división de tus datos escalados en conjuntos de entrenamiento y prueba. Los conjuntos resultantes son `X_train`, `X_test`, `y_train` y `y_test`, que se utilizan para entrenar y evaluar modelos de aprendizaje automático. El uso de una semilla aleatoria (random_state) garantiza que la división sea consistente cada vez que se ejecute el código.

 ## **<font color="DarkBlue">Ajuste del modelo con sklearn</font>**

<p align="justify">
La clase <code>LinearRegression</code> de <code>scikit-learn</code> proporciona una interfaz sencilla y eficiente para ajustar modelos de regresión lineal y realizar predicciones. Es ampliamente utilizada en problemas de regresión en los que se busca modelar y predecir una variable dependiente a partir de variables independientes mediante una relación lineal.
<br>
<br>
En este caso, se intenta predecir el valor del <code>Gasto_anual</code> en función de las variables explicativas <code>Tiempo_sesión</code>, <code>Tiempo_app</code>,	<code>Tiempo_web</code> y <code>Años_miembro</code>.
<br>
<br>
👀 Problema de regresión lineal múltiple.

In [ ]:
model_1 = LinearRegression().fit(X_train, y_train)
print("")
print(f"intercept = {model_1.intercept_}")
print(f"coef = {model_1.coef_}")


intercept = 499.14110628577623
coef = [25.14058939 38.60261634  0.77873252 61.59498109]


$Gasto \ anual = 499.1+25.1 × Tiempo \ sesión + 38.6 × Tiempo \ app + 0.8 × Tiempo \ web + 61.6 × Años\ miembro$

**Interpretación de los coeficientes:**

Manteniendo las otras características fijas,
+ un incremento de 1 unidad en `Tiempo_sesión` está asociado con un incremento de 25.14 en `Gasto_anual`,
+ un incremento de 1 unidad en `Tiempo_app` está asociado con un incremento de 38.6 en `Gasto_anual`,
+ un incremento de 1 unidad en `Tiempo_web` está asociado con un incremento de 0.77 en `Gasto_anual`y
+ un incremento de 1 unidad en `Años_miembro` está asociado con un incremento de 61.59 en `Gasto_anual`.

<p align="justify">
En este fragmento de código, estás ajustando un modelo de regresión lineal a tus datos de entrenamiento y luego imprimiendo los coeficientes del modelo resultante. A continuación, se detalla cada parte:

- `model_1 = LinearRegression().fit(X_train, y_train)`: Aquí, estás creando una instancia del modelo de regresión lineal utilizando la clase `LinearRegression` de scikit-learn. Luego, estás ajustando (entrenando) el modelo utilizando los conjuntos de entrenamiento "X_train" (características) y "y_train" (variable dependiente). Esto significa que el modelo está encontrando la línea de regresión que mejor se ajusta a tus datos de entrenamiento.

- `print("")`: Esto simplemente imprime una línea en blanco en la consola, para separar visualmente los resultados anteriores.

- `print(f"intercept = {model_1.intercept_}")`: Aquí estás imprimiendo el término de intersección del modelo, que se encuentra en el atributo `intercept_` del objeto `model_1`. Este término representa el valor predicho de la variable dependiente cuando todas las características son cero.

- `print(f"coef = {model_1.coef_}")`: Aquí estás imprimiendo los coeficientes de las características del modelo. Estos coeficientes se almacenan en el atributo `coef_` del objeto `model_1`. Los coeficientes indican cómo cambia la variable dependiente en relación con cambios en las características, manteniendo constantes las demás variables.

<p align="justify">
En resumen, este código ajusta un modelo de regresión lineal a los datos de entrenamiento escalados. Luego, imprime el término de intersección (término constante) y los coeficientes de las características del modelo. Estos valores son importantes para entender cómo el modelo está realizando sus predicciones basadas en las características proporcionadas.

## **<font color="DarkBlue">Predicción y evaluación del modelo**

In [ ]:
prediction = model_1.predict(X_test)
prediction[:5]

array([451.15603635, 449.76991439, 565.6671197 , 506.74017581,
       389.26917501])

In [ ]:
tabla = pd.DataFrame({"Prediccion":prediction,
                      "Real":y_test,
                      "Residuos": (y_test-prediction),
                      })
tabla.head()

Prediccion        Real   Residuos
229  451.156036  436.283498 -14.872538
337  449.769914  440.002748  -9.767167
327  565.667120  557.252687  -8.414433
416  506.740176  511.038786   4.298610
306  389.269175  387.534716  -1.734459

In [ ]:
model_1.score(X_test, y_test).round(4)

0.985

<p align="justify">
En este fragmento de código, estás ajustando un modelo de regresión lineal a tus datos de entrenamiento y luego imprimiendo los coeficientes del modelo resultante. A continuación, se detalla cada parte:

- `model_1 = LinearRegression().fit(X_train, y_train)`: Aquí, estás creando una instancia del modelo de regresión lineal utilizando la clase `LinearRegression` de scikit-learn. Luego, estás ajustando (entrenando) el modelo utilizando los conjuntos de entrenamiento "X_train" (características) y "y_train" (variable dependiente). Esto significa que el modelo está encontrando la línea de regresión que mejor se ajusta a tus datos de entrenamiento.

- `print("")`: Esto simplemente imprime una línea en blanco en la consola, para separar visualmente los resultados anteriores.

- `print(f"intercept = {model_1.intercept_}")`: Aquí estás imprimiendo el término de intersección del modelo, que se encuentra en el atributo `intercept_` del objeto `model_1`. Este término representa el valor predicho de la variable dependiente cuando todas las características son cero.

- `print(f"coef = {model_1.coef_}")`: Aquí estás imprimiendo los coeficientes de las características del modelo. Estos coeficientes se almacenan en el atributo `coef_` del objeto `model_1`. Los coeficientes indican cómo cambia la variable dependiente en relación con cambios en las características, manteniendo constantes las demás variables.

<p align="justify">
En resumen, este código ajusta un modelo de regresión lineal a los datos de entrenamiento escalados. Luego, imprime el término de intersección (término constante) y los coeficientes de las características del modelo. Estos valores son importantes para entender cómo el modelo está realizando sus predicciones basadas en las características proporcionadas.

## **<font color="DarkBlue">Creación del Pipeline**

In [ ]:
model = make_pipeline(StandardScaler(), LinearRegression())
model

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearregression', LinearRegression())])

<p align="justify">
En este fragmento de código, estás creando un modelo de regresión lineal utilizando el concepto de un pipeline en scikit-learn. Aquí está la explicación:

- `StandardScaler()`: Esta es una instancia del escalador estándar que estandarizará los datos antes de pasarlos al modelo de regresión lineal. El escalador estándar transforma los datos para que tengan una media de 0 y una desviación estándar de 1.

- `LinearRegression()`: Esta es una instancia del modelo de regresión lineal. Será el segundo paso en el pipeline después del escalado.

- `make_pipeline(StandardScaler(), LinearRegression())`: Aquí estás utilizando la función `make_pipeline()` de scikit-learn para crear un pipeline que combina el escalador estándar y el modelo de regresión lineal en ese orden. Los datos pasarán primero por el escalador y luego por el modelo de regresión lineal.

- `model`: Después de crear el pipeline, se almacena en la variable `model`.

<p align="justify">
En resumen, estás creando un modelo de regresión lineal utilizando un pipeline que primero estandariza los datos utilizando el escalador estándar y luego ajusta un modelo de regresión lineal a los datos escalados. <mark>El uso de un pipeline como este puede simplificar el proceso de ajustar y evaluar modelos, ya que combina varios pasos en un solo objeto</mark>.

 # **<font color="DarkBlue"> 💗 Los esperamos...</font>**